plugins/store/guest-js/index.ts

```diff
@@ -8,50 +8,211 @@ import { invoke, Resource } from '@tauri-apps/api/core'
 
 interface ChangePayload<T> {
   path: string
+  resourceId?: number
   key: string
-  value: T | null
+  value: T
+  exists: boolean
 }
 
 /**
  * Options to create a store
  */
 export type StoreOptions = {
   /**
-   * Auto save on modification with debounce duration in milliseconds
+   * Auto save on modification with debounce duration in milliseconds, it's 100ms by default, pass in `false` to disable it
    */
-  autoSave?: boolean
+  autoSave?: boolean | number
+  /**
+   * Name of a serialize function registered in the rust side plugin builder
+   */
+  serializeFnName?: string
+  /**
+   * Name of a deserialize function registered in the rust side plugin builder
+   */
+  deserializeFnName?: string
+}
+
+/**
+ * @param path: Path to save the store in `app_data_dir`
+ * @param options: Store configuration options
+ *
+ * @throws If a store at that path already exists
+ */
+export async function createStore(
+  path: string,
+  options?: StoreOptions
+): Promise<Store> {
+  return await Store.createStore(path, options)
 }
 
 /**
+ * Create a new Store or get the existing store with the path
+ *
  * @param path: Path to save the store in `app_data_dir`
  * @param options: Store configuration options
  */
-export async function createStore(path: string, options?: StoreOptions) {
-  const resourceId = await invoke<number>('plugin:store|create_store', {
-    path,
-    ...options
-  })
-  return new Store(resourceId, path)
+export async function createOrExistingStore(
+  path: string,
+  options?: StoreOptions
+): Promise<Store> {
+  return await Store.createOrExistingStore(path, options)
+}
+
+/**
+ * @param path: Path of the store in the rust side
+ */
+export async function getStore(path: string): Promise<Store | undefined> {
+  return await Store.getStore(path)
 }
 
 /**
  * A lazy loaded key-value store persisted by the backend layer.
  */
-export class Store extends Resource {
+export class LazyStore implements IStore {
+  private _store?: Promise<Store>
+
+  private get store(): Promise<Store> {
+    if (!this._store) {
+      this._store = createOrExistingStore(this.path, this.options)
+    }
+    return this._store
+  }
+
+  /**
+   * Note that the options are not applied if someone else already created the store
+   * @param path: Path to save the store in `app_data_dir`
+   * @param options: Store configuration options
+   */
   constructor(
-    rid: number,
-    private readonly path: string
-  ) {
+    private readonly path: string,
+    private readonly options?: StoreOptions
+  ) {}
+
+  /**
+   * Init/load the store if it's not already
+   */
+  async init(): Promise<void> {
+    await this.store
+  }
+
+  async set(key: string, value: unknown): Promise<void> {
```

Comment:
They share the same docs, why do we want to duplicate that?

Reply:
My main concern is hover documentation and documentation on our website, also `IStore` shouldn't be public as well, which could mess up the target documentation cases (didn't test).

Reply:
Hover documentation works fine for me, not sure the docs on the website though, I think `IStore` needs to be public so people could use the two stores interchangeably

Reply:
but why would user need to use two types of stores interchangeably?

Reply:
Well, you might want to put the store inside a context (in react/solidjs/etc term) and type it as `IStore` instead of a specific store, if you really don't like this we can not to expose it for now and see if anyone requests for it

Reply:
but why would I store in a context with `IStore`, if I am using `LazyStore` why not just store it as `LazyStore`. I fail to see a real use-case here tbh, I'd recommend removing.

Reply:
Also realistically speaking you're gonna use one or the other but never both. And it is part of why I don't really like the addition of `LazyStore`.

Reply:
I honestly think the better usage is through `createStore` and `Store`. `LazyStore` feels kinda magical and it also delays the errors emitted during initialization of the store until a method is used which will confuse the user as to why this method is throwing this unrelated error.

Reply:
Just realized that `IStore` was never exposed

> Also realistically speaking you're gonna use one or the other but never both

That's fair

> I honestly think the better usage is through createStore and Store.
> 
> LazyStore feels kinda magical and it also delays the errors emitted during initialization of the store until a method is used which will confuse the user as to why this method is throwing this unrelated error.

I thought this way as well first when I made the last rework, but then I realized that in UI usages, something like `localStorage` is the easiest to use and despite the downsides, not many people would choose indexeddb over it in simple use cases, I think this is somewhat true for our case as well